src/state/slices/txHistorySlice/txHistorySlice.ts

```diff
@@ -144,6 +151,64 @@ export const txHistory = createSlice({
       state.status = payload
     },
     onMessage: (txState, { payload }: TxMessage) =>
-      updateOrInsert(txState, payload.message, payload.accountSpecifier)
+      updateOrInsert(txState, payload.message, payload.accountSpecifier),
+    upsertTxs: (txState, { payload }) => {
+      // TODO: implementation
+      for (const tx of payload.txs) {
+        updateOrInsert(txState, tx, payload.accountSpecifier)
+      }
+    }
   }
 })
+
+type AllTxHistoryArgs = { accountSpecifierMap: AccountSpecifierMap }
+
+export const txHistoryApi = createApi({
+  reducerPath: 'marketApi',
+  // not actually used, only used to satisfy createApi, we use a custom queryFn
+  baseQuery: fetchBaseQuery({ baseUrl: '/' }),
+  // refetch if network connection is dropped, useful for mobile
+  refetchOnReconnect: true,
+  endpoints: build => ({
+    getAllTxHistory: build.query<chainAdapters.Transaction<ChainTypes>[], AllTxHistoryArgs>({
+      queryFn: async ({ accountSpecifierMap }, { dispatch }) => {
+        try {
+          if (isEmpty(accountSpecifierMap))
+            throw new Error('getAllTxHistory: No accounts given to get transaction history')
+
+          let txs: chainAdapters.Transaction<ChainTypes>[] = []
+          const chainAdapters = getChainAdapters()
+          const [CAIP2, pubkey] = Object.entries(accountSpecifierMap)[0] as [CAIP2, string]
+          const accountSpecifier = `${CAIP2}:${pubkey}`
+          const { chain } = caip2.fromCAIP2(CAIP2)
+          const adapter = chainAdapters.byChain(chain)
+          let currentCursor: string = ''
+          const pageSize = 100
+          do {
+            const { cursor: _cursor, transactions } = await adapter.getTxHistory({
+              cursor: currentCursor,
+              pubkey,
+              pageSize
+            })
+            currentCursor = _cursor
+            txs = [...txs, ...transactions]
+          } while (currentCursor)
+          dispatch(txHistory.actions.upsertTxs({ txs, accountSpecifier }))
+          return { data: txs }
+        } catch (err) {
+          if (err instanceof Error) {
+            const data = err.message
+            const error = { data, status: 400 }
+            return { error }
+          }
+          return {
+            error: {
+              data: 'getAllTxHistory: An error occurred fetching all tx history',
+              status: 500
+            }
+          }
```

Comment:
If anyone has any better ideas here for the error handling, I'm all ears. I don't like the idea of setting error status on the client, but RTK Query requires status on the returned error.

Reply:
Yeah, that's also a very usual react way to do things, many data-fetching libraries return a tuple like `{ error, data, loading }`, so that's not shocking to me